miio/airpurifier_miot.py

```diff
@@ -0,0 +1,405 @@
+                "average_aqi": {"siid": 13, "piid": 2},
+                # RFID (siid=14)
+                "filter_rfid_tag": {"siid": 14, "piid": 1},
+                "filter_rfid_product_id": {"siid": 14, "piid": 3},
+                # Other (siid=15)
+                "app_extra": {"siid": 15, "piid": 1},
+            },
+            ip,
+            token,
+            start_id,
+            debug,
+            lazy_discover,
+        )
+
+    @command(
+        default_output=format_output(
+            "",
+            "Power: {result.power}\n"
+            "AQI: {result.aqi} μg/m³\n"
+            "Average AQI: {result.average_aqi} μg/m³\n"
+            "Humidity: {result.humidity} %\n"
+            "Temperature: {result.temperature} °C\n"
+            "Fan Level: {result.fan_level}\n"
+            "Mode: {result.mode}\n"
+            "LED: {result.led}\n"
+            "LED brightness: {result.led_brightness}\n"
+            "Buzzer: {result.buzzer}\n"
+            "Child lock: {result.child_lock}\n"
+            "Favorite level: {result.favorite_level}\n"
+            "Filter life remaining: {result.filter_life_remaining} %\n"
+            "Filter hours used: {result.filter_hours_used}\n"
+            "Use time: {result.use_time} s\n"
+            "Purify volume: {result.purify_volume} m³\n"
+            "Motor speed: {result.motor_speed} rpm\n"
+            "Filter RFID product id: {result.filter_rfid_product_id}\n"
+            "Filter RFID tag: {result.filter_rfid_tag}\n"
+            "Filter type: {result.filter_type}\n",
+        )
+    )
+    def status(self) -> AirPurifierMiotStatus:
+        """Retrieve properties."""
+
+        return AirPurifierMiotStatus(
+            {prop["did"]: prop["value"] for prop in self.get_properties()}
+        )
+
+    @command(default_output=format_output("Powering on"))
+    def on(self):
+        """Power on."""
+        return self.set_property("power", True)
+
+    @command(default_output=format_output("Powering off"))
+    def off(self):
+        """Power off."""
+        return self.set_property("power", False)
+
+    @command(
+        click.argument("level", type=int),
+        default_output=format_output("Setting fan level to '{level}'"),
+    )
+    def set_fan_level(self, level: int):
+        """Set fan level."""
+        if level < 1 or level > 3:
+            raise AirPurifierMiotException("Invalid fan level: %s" % level)
+        return self.set_property("fan_level", level)
+
+    @command(
+        click.argument("mode", type=EnumType(OperationMode, False)),
+        default_output=format_output("Setting mode to '{mode.value}'"),
+    )
+    def set_mode(self, mode: OperationMode):
+        """Set mode."""
+        return self.set_property("mode", mode.value)
+
+    @command(
+        click.argument("level", type=int),
+        default_output=format_output("Setting favorite level to {level}"),
+    )
+    def set_favorite_level(self, level: int):
```

Comment:
Just curious if you've tried out piid: 9 (motor1-set-speed): (value-range: [0, 3000, 1]). Can it be used to set motor rpm speed directly instead of using favorite_level steps? We can have a set_motor_speed function then.

Would be useful for better speed control with hysteresis instead of jumps in motor speed. Probably better for the motor, too.

And anyone using the library can map it better to their apps.

Reply:
@mouth4war Thanks, great suggestion. I tried the `{siid:10, piid:9}` which didn't work for me:
```
$ python miio/cli.py airpurifiermiot --ip $AIR_IP --token $TOKEN set_motor_speed 1000
Setting motor speed '1000' rpm
[{'did': 'set_motor_speed', 'siid': 10, 'piid': 9, 'code': -4003}]
```
however, it turned out that `{siid:10, piid:7}` works fine -- at least for values between 230 and 2200 (rather than 2300 as per specs):
```
$ python miio/cli.py airpurifiermiot --ip $AIR_IP --token $TOKEN set_favorite_rpm 2300
Setting favorite motor speed '2300' rpm
[{'did': 'set_favorite_rpm', 'siid': 10, 'piid': 7, 'code': -4005}]
$ python miio/cli.py airpurifiermiot --ip $AIR_IP --token $TOKEN set_favorite_rpm 2210
Setting favorite motor speed '2210' rpm
[{'did': 'set_favorite_rpm', 'siid': 10, 'piid': 7, 'code': -4005}]
$ python miio/cli.py airpurifiermiot --ip $AIR_IP --token $TOKEN set_favorite_rpm 290
Setting favorite motor speed '290' rpm
[{'did': 'set_favorite_rpm', 'siid': 10, 'piid': 7, 'code': -4005}]
$ python miio/cli.py airpurifiermiot --ip $AIR_IP --token $TOKEN set_favorite_rpm 300
Setting favorite motor speed '300' rpm
[{'did': 'set_favorite_rpm', 'siid': 10, 'piid': 7, 'code': 0}]
$ python miio/cli.py airpurifiermiot --ip $AIR_IP --token $TOKEN set_favorite_rpm 2200
Setting favorite motor speed '2200' rpm
[{'did': 'set_favorite_rpm', 'siid': 10, 'piid': 7, 'code': 0}]
```

Pushed additional 2 commits. Note: please ignore test failures for now; I'm trying to get https://github.com/rytilahti/python-miio/pull/592 reviewed first and the i'll follow up with this.